### PR TITLE
Add rename file action

### PR DESCRIPTION
**Pull request by PCSwingle**

### CODE REVIEW

File: mentat/code_change_display.py
Comment (biobootloader) [inline mentat/code_change_display.py:68]:
```suggestion
            return colored(f"\nRename: {file_name} -> {code_change.name}", color="yellow")
```

**Review by biobootloader — COMMENTED** [review]
Nice!

I tried it out to see what would happen if I asked for a change and a rename in the same file, and I end up not seeing the rename change at all? See result and log below. In this case what I think we should do is display "Rename x.py -> y.py" above the list of changes (if any) for that file (aka, that would replace the usual line showing which file is being changed above other changes). 

image
image

**Comment by PCSwingle** [comment]
The model hasn't given the rename file action by the time we've streamed the file name for the previous response; additionally, since the model's responses are applied in order, I think it's best to keep the change files the same as they are. I'm going to fix it to print the rename file action, but not change the other actions at all.

**Comment by biobootloader** [comment]
@PCSwingle 

> The model hasn't given the rename file action by the time we've streamed the file name for the previous response

ah...

> additionally, since the model's responses are applied in order

within a file we sort changes, but you mean the order that we change files?

**Comment by biobootloader** [comment]
I tried running this again and it still doesn't show me the rename? And then I applied all changes and it crashed after deleting the file (but not making a new one):

image
image

**Comment by PCSwingle** [comment]
Just tried your prompt and it worked? Maybe it got fixed when I rebased on main (since there were a lot of path changes and things that I had to change). If you have the logs from that run still though I can see if I can reproduce it

**Review by biobootloader — APPROVED** [review]
Works for me now! Ready to merge after fixing conflict

**Comment by PCSwingle** [comment]
@waydegg just finished fixing merge conflicts; would love to get your feedback on this PR before I merge it!

File: mentat/app.py
Comment (waydegg) [inline mentat/app.py:208]: What edge cases have you discovered/thought about so far?

One that comes to mind is a file being renamed and imports not being updated accordingly.

**Review by waydegg — APPROVED** [review]
Left a few comments but I don't think there's anything preventing these changes to be merged. The updated system prompt seems to hold up for pretty complex instructions I've thrown at it form just manually testing so great work!

File: mentat/app.py
Comment (PCSwingle) [reply mentat/app.py:208]: I was thinking more edge cases in how we handle it; the big one being that if we don't rename the file, then any changes the model suggests that try to change the renamed file will be trying to change a non-existent file. It gets worse if the model is renaming multiple files to similar names, or one file twice, and so on.

File: mentat/code_context.py
Comment (PCSwingle) [reply mentat/code_context.py:89]: Well the real reason was because we were actually just putting a relative Path in there (when it was supposed to be an absolute path!); but yeah, the reason I use realpath instead of abspath is because of symlinks

File: mentat/code_file_manager.py
Comment (PCSwingle) [reply mentat/code_file_manager.py:91]: This is actually necessary; abspath appends the CWD to the beginning of the path, which might not necessarily be the current git repo (it could be a subdirectory). This does remind me though, I don't think we have any tests where we run mentat from a subdiretory of the git repo it's actually in; we should probably think about adding some of those!

Comment (biobootloader) [reply mentat/code_file_manager.py:91]: > I don't think we have any tests where we run mentat from a subdiretory of the git repo it's actually in; we should probably think about adding some of those!

@PCSwingle can you add an issue for this

**Pull request merged**
